### ISSUE TITLE
Unable to install Microsoft.VisualStudio.Workload.VCTools from LTSC2019 base image

### ISSUE THREAD

**Issue by perezd** [microsoft/dotnet-framework-docker#847]
### Describe the Bug

using vs_buildtools.exe to install `Microsoft.VisualStudio.Workload.VCTools` no longer works as it used to.

### Steps to Reproduce

Dockerfile example:
```
FROM mcr.microsoft.com/dotnet/framework/sdk:4.8-windowsservercore-ltsc2019

ADD https://aka.ms/vs/16/pre/vs_buildtools.exe C:\vs_buildtools.exe
RUN Start-Process -Wait -FilePath C:\vs_buildtools.exe `
 -ArgumentList @('--quiet --wait --norestart --nocache --add Microsoft.VisualStudio.Workload.VCTools --includeOptional --includeRecommended')
```

### Other Information

This used to work, I did a successful image build (Sep 9, 2021, 5:46:42 PM) where the workload was properly installed, and trying again today changing nothing about the Dockerfile it is failing.

### Output of `docker version`

```
Client: Mirantis Container Runtime
 Version:           19.03.18
 API version:       1.40
 Go version:        go1.13.15
 Git commit:        5b7ee53
 Built:             08/19/2021 16:42:56
 OS/Arch:           windows/amd64
 Experimental:      false

Server: Mirantis Container Runtime
 Engine:
  Version:          19.03.18
  API version:      1.40 (minimum version 1.24)
  Go version:       go1.13.15
  Git commit:       aed55fd091
  Built:            08/19/2021 16:41:33
  OS/Arch:          windows/amd64
  Experimental:     false
```

### Output of `docker info`

```
Client:
 Debug Mode: false
 Plugins:
  cluster: Manage Mirantis Container Cloud clusters (Mirantis Inc., v1.9.0)

Server:
 Containers: 9
  Running: 1
  Paused: 0
  Stopped: 8
 Images: 25
 Server Version: 19.03.18
 Storage Driver: windowsfilter
  Windows:
 Logging Driver: json-file
 Plugins:
  Volume: local
  Network: ics internal l2bridge l2tunnel nat null overlay private transparent
  Log: awslogs etwlogs fluentd gcplogs gelf json-file local logentries splunk syslog
 Swarm: inactive
 Default Isolation: process
 Kernel Version: 10.0 17763 (17763.1.amd64fre.rs5_release.180914-1434)
 Operating System: Windows Server 2019 Datacenter Version 1809 (OS Build 17763.2237)
 OSType: windows
 Architecture: x86_64
 CPUs: 16
 Total Memory: 64GiB
 Name: pzd-windows-2019-2
 ID: JH44:YSBL:PYB2:CW27:R7WI:6XNM:IB5V:EEXU:BSOI:PHH5:OI2Y:TGSW
 Docker Root Dir: C:\ProgramData\docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by mthalman** [comment]
Can you post what the digest value is for the mcr.microsoft.com/dotnet/framework/sdk:4.8-windowsservercore-ltsc2019 image?

**Comment by perezd** [comment]
The one that is not working is: `mcr.microsoft.com/dotnet/framework/sdk@sha256:82047a37202950bc5fcb6f47938f9b2871f94bde9bddf7f574a859289e15103c`

**Comment by mthalman** [comment]
Can you be more specific about what doesn't work? Is there an error?

Note that we _just very recently_ published SDK images that now contain [tools for Visual Studio 2022](https://github.com/microsoft/dotnet-framework-docker/issues/848). Note sure if you were making use of that since they were just published.

**Comment by mthalman** [comment]
Ok, that digest indicates it is not the image we just published today.

**Comment by perezd** [comment]
> Can you be more specific about what doesn't work? Is there an error?
> 
> Note that we _just very recently_ published SDK images that now contain [tools for Visual Studio 2022](https://github.com/microsoft/dotnet-framework-docker/issues/848). Note sure if you were making use of that since they were just published.

I wish I had more info to give you, it just silently exits and I do not see the expected folders in C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC.

Here's also the manifest of an image where it used to work:
```
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
   "config": {
      "mediaType": "application/vnd.docker.container.image.v1+json",
      "size": 13208,
      "digest": "sha256:ed613b46c13378e8d9cf202e71d41755b54b3f6e94a9ab32a58b5d92167a6566"
   },
   "layers": [
      {
         "mediaType": "application/vnd.docker.image.rootfs.foreign.diff.tar.gzip",
         "size": 1718332879,
         "digest": "sha256:4612f6d0b889cad0ed0292fae3a0b0c8a9e49aff6dea8eb049b2386d9b07986f",
         "urls": [
            "https://mcr.microsoft.com/v2/windows/servercore/blobs/sha256:4612f6d0b889cad0ed0292fae3a0b0c8a9e49aff6dea8eb049b2386d9b07986f"
         ]
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.foreign.diff.tar.gzip",
         "size": 967665054,
         "digest": "sha256:c67ded6868b61d392a0c096f911563fd6bc0bc3ed4fe401d077b3718a1b0cdaf",
         "urls": [
            "https://mcr.microsoft.com/v2/windows/servercore/blobs/sha256:c67ded6868b61d392a0c096f911563fd6bc0bc3ed4fe401d077b3718a1b0cdaf"
         ]
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 1279,
         "digest": "sha256:acfd20817b940bc3b788cbe8d03147780d18147788968e3661892255e70de155"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 887596480,
         "digest": "sha256:19b521ca6db2368ac222e93640f3dc26744b042516f2bc6f92bdb9273f9dc8af"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 1274,
         "digest": "sha256:bc3a76c4ed08e7b88288d169e3498362e90d213e0dc4ca8a8c05b4e38b434ba6"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 8544626,
         "digest": "sha256:1ae387f5f555ade511aa913db8db8b96debadf729df4af13ca996975f1d311c3"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 1294,
         "digest": "sha256:3d091581326d1587bf1b0489c5ad8844520b11ac6f5c5af0abadfb742a179384"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 1085083499,
         "digest": "sha256:2339606413361e999a00b3beda519e6e36c57e2621b3ac8bd984f4e34ffb9d79"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 838427,
         "digest": "sha256:77ae1466f1658ca81c16b7566402421af17883a3cb0609839fc3e4807330ce8b"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 1270,
         "digest": "sha256:28d33e51bb8541da4c8c79d862562e92f32bee6f7ed3aec5c5a6e9b291f828f7"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 598243,
         "digest": "sha256:73362e7177010dac9be7e3c4af6dff70ea234f91dd358bd527ffa1ad97744144"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 167608235,
         "digest": "sha256:5f867d39933f7240fdb167278ffb37e9107a5acc81d140685a7f3637bc0e8d59"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 1269,
         "digest": "sha256:2df7f424f187731273011a2013cd08a8940bf8c749f607ac406e34e2aefd0912"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 41907282,
         "digest": "sha256:9280b64d618ce8fb11af6ad260ae2dd4d31cc9b472f8b22dc33c5f57e0894c15"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 73922565,
         "digest": "sha256:9f45bbee5588d11376f48bf1b4c5786a42e4a415f5be0fdd4af2b1fed996489e"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 1347634,
         "digest": "sha256:d2189c03ee8ba7aeecededc26c9064f4f0d40aeb7e398b474432dd850d80c594"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 1693,
         "digest": "sha256:83c3260aa7d03b7c1d00f19c293d0e7f0c56920f472e90784157bdf8a1a2221f"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 1566,
         "digest": "sha256:4a1964e9efc67d87e4d778692ccd47a77082ec109d6b65b845ac009e5ba46edc"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 143608091,
         "digest": "sha256:d1cc5282912aeb2bd4b2d0211c40d5af05e5b2849f8a118ea3009a4b859d47fd"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 2021709525,
         "digest": "sha256:13636e034ca4e9fb8db46ae570205e484081ae6c81a9fdceace5099fbeb3d88d"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 1302,
         "digest": "sha256:4971133820ba409c6272f936c321cf93dcdcf7a19698798903a7ef59e89e276d"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 1297,
         "digest": "sha256:e6b57187370afcc5b8d90a4b4cf16d1d839fadeff7e4e49116407f6adf9f3841"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 1299,
         "digest": "sha256:87b79ea82767abaf206cd4a463701127c9b8245032b907fb371ab6aaeb0b680f"
      }
   ]
}
```

**Comment by perezd** [comment]
If you have a way to squeeze more useful diagnostics out of vs_buildtools.exe lmk, I can try!

**Comment by perezd** [comment]
FWIW, I was using a cached version of the vs_buildtools.exe that was identical to the previous runs that did succeed, I just did that to make reproing easier.

I've created that file at `C:\ProgramData\docker\config\daemon.json`, (set it to size=500GB) and restarted the docker engine service, followed by a system prune -a. 

Next, I've done a build of the Dockerfile.

I see this in the following file `t C:\Users\ContainerAdministrator\AppData\Local\Temp\dd_installer_elevated_20211109224001.log`
```
[0798:0017][2021-11-09T22:40:12] Set client executable and run once arguments [setupExecutablePath: "C:\Program Files (x86)\Microsoft Visual Studio\Installer\setup.exe", setupExecutableArguments: "resume --installPat
h "C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools" --runOnce --installSessionId e331d88b-724a-497b-a18e-6088bde03a48 --quiet"]
[0798:0017][2021-11-09T22:40:12] Error 0x57: Install operation failed with error: The parameter is incorrect
   at Microsoft.VisualStudio.Setup.Services.ConsoleLogger.Attach(Int32 processId)
   at Microsoft.VisualStudio.Setup.Engine.AggregateLogger(ILogger clientLogger)
   at Microsoft.VisualStudio.Setup.Engine.Initialize()
   at Microsoft.VisualStudio.Setup.Services.Installer.InstallerBase.<>c__DisplayClass5_0.<<-ctor>g__initializeAsync|3>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Setup.AsyncLazy`1.<GetValueAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Setup.Services.Installer.InstallerBase.<InitializeAsync>d__50.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Setup.Services.Installer.InstallerBase.<InvokeAsync>d__16.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Setup.Installer.Installer.InstallerService.<Execute>d__26.MoveNext()
[0798:0001][2021-11-09T22:40:12] Closing the installer with exit code 0```

**Comment by mthalman** [comment]
Do any of the other log files in the temp folder have any content that provides more detail?

**Comment by mthalman** [comment]
One thing I noticed is that you're looking for the VC directory at `C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC` but you're installing the preview version of VS from `https://aka.ms/vs/16/pre/vs_buildtools.exe`. Preview versions would not get installed by default to the `BuildTools` directory but rather to the `Preview` directory. So the path you should be looking for is `C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC`.

Is there any particular reason you're targeting the preview version instead of the release version (i.e. https://aka.ms/vs/16/release/vs_buildtools.exe)?

In any case, if you're going to be using the latest SDK image, it has been updated to VS 2022 so you'll want to change the version to VS 17.0 (`https://aka.ms/vs/17/release/vs_buildtools.exe`) when targeting the latest SDK image.

**Comment by perezd** [comment]
So a few interesting things:

1. I'm actually using 16 release (I think), but it's copied locally from disk, not loaded w/ aka.ms, I just did that to make life easier for repro. It used to work as expected, literally nothing changed on my end.
2. It forever has been writing to a `2019` folder and suddenly (again, same executable from before) it's writing to `2022`.
3. I zipped up the whole temp folder from within the container for you to check out:
[Logs.zip](https://github.com/microsoft/dotnet-framework-docker/files/7514606/Logs.zip)

**Comment by perezd** [comment]
Found version info of the actual executable I'm using:
![Screen Shot 2021-11-10 at 9 15 19 AM](https://user-images.githubusercontent.com/20598/141160852-99c05c49-fef7-43c8-8f75-cd4cb438285c.png)

**Comment by mthalman** [comment]
I don't know where the `vs_buildtools.exe` file came from that you've got cached so it's really hard for me to say anything about that.

What I've tried is building the Dockerfile you've provided in a Docker configuration that sets the storage size to 100GB. For the base image, I'm using the image associated with the digest you provided. When I do that, the build succeeds and the VC directory exists.

> * It forever has been writing to a `2019` folder and suddenly (again, same executable from before) it's writing to `2022`. (OR, this executable is writing nothing, and it was present in the base image that got updated underneath me)

That means you've pulled down a newer version of the SDK image. As I've said, the SDK has just yesterday been updated to have VS 2022 tools installed: https://github.com/microsoft/dotnet-framework-docker/issues/848. So because now you're using a different base image from what you were originally, we need to re-evaluate this. I also got the same error when pulling the latest SDK image and installing with `https://aka.ms/vs/16/pre/vs_buildtools.exe`. But that's not anything I would recommend because you're mixing VS versions in the container which leads to unnecessary bloat. I don't know the exact cause of the error but I wouldn't recommend this approach anyway. Instead, you should be using the `https://aka.ms/vs/17/release/vs_buildtools.exe` URL when targeting the latest SDK image. I've done that locally and it installs fine (assuming that the Docker daemon storage size config changes have been made).

In other words, this works for me:

```Dockerfile
FROM mcr.microsoft.com/dotnet/framework/sdk:4.8-windowsservercore-ltsc2019

ADD https://aka.ms/vs/17/release/vs_buildtools.exe C:\vs_buildtools.exe
RUN Start-Process -Wait -FilePath C:\vs_buildtools.exe `
 -ArgumentList @('--quiet --wait --norestart --nocache --add Microsoft.VisualStudio.Workload.VCTools --includeOptional --includeRecommended')
```

```console
$ docker build -t test --pull .
```

**Comment by perezd** [comment]
OK so I've updated my vs_buildtools.exe to match 17/release and it seems to work now. I have no idea how the other one get messed up, but I guess that it was incompatible w/ the changes to the base image in some weird way. Thanks for debugging w/ me!
